alf/algorithms/distributed_off_policy_algorithm.py

```diff
@@ -662,6 +677,35 @@ def observe_for_replay(self, exp: Experience):
         Every time we make sure a full episode is sent to the same DDP rank, if
         multi-gpu training is enabled on the trainer.
         """
+        # Get the current worker id to send the exp to
+        worker_id = f'worker-{self._current_worker}'
+        self._num_exps += 1
+        episode_end = ((self._episode_length <= 0 and bool(exp.is_last()))
+                       or (self._num_exps % self._episode_length == 0))
+
+        if self._is_first_step:
+            # When the unroller has a ``max_episode_length``, we need to correctly
+            # set first time step type.
+            if not episode_end:
+                # In rare cases, the first step is also the last step, we don't
+                # overwrite LAST to FIRST
+                exp = alf.nest.set_field(
+                    exp, 'time_step.step_type',
+                    torch.tensor([StepType.FIRST], dtype=torch.int32))
+            self._is_first_step = False
+
+        if episode_end:
+            # One episode finishes; move to the next worker
+            # We need to make sure a whole episode is always sent to the same
+            # worker so that the temporal information is preserved.
+            exp = alf.nest.set_field(
```

Comment:
In the case of a single trainer workers, we don't need to change the step type to LAST.

Reply:
> In the case of a single trainer workers, we don't need to change the step type to LAST.

If there are multiple unrollers, we still need to set LAST. But it's not straightforward for an unroller to know if there is any other unroller, unless via the trainer. So for simplicity, here we always set LAST.